File: src/tool.cc
#include <unistd.h>

#include <cerrno>
#include <cstdlib>
#include <cstring>
#include <filesystem>
#include <iostream>
#include <optional>
#include <string>
#include <string_view>
#include <vector>


constexpr std::string_view ANSI_CSI{"\033["};
constexpr char ANSI_SEPERATOR{';'};
constexpr char ANSI_SGR{'m'};
constexpr std::string_view ANSI_RESET{"0"};
constexpr std::string_view ANSI_BOLD{"1"};
constexpr std::string_view ANSI_RED{"31"};

constexpr std::string JAVA_HOME{"JAVA_HOME"};
constexpr std::string_view BIN{"bin"};


template<typename... Args>
void error(Args&&... args) {
	std::cerr << ::ANSI_CSI << ::ANSI_RED << ::ANSI_SGR;
	((std::cerr << args << ' '), ...);
	std::cerr << ::ANSI_CSI << ::ANSI_RESET << ::ANSI_SGR;
	std::cerr << '\n';
}

template<typename... Args>
void java_home_error(Args&&... args) {
	std::cerr << ::ANSI_CSI << ::ANSI_RED << ::ANSI_SGR << '\'';
	std::cerr << ::ANSI_CSI << ::ANSI_BOLD << ::ANSI_SEPERATOR << ::ANSI_RED << ::ANSI_SGR << ::JAVA_HOME;
	std::cerr << ::ANSI_CSI << ::ANSI_RED << ::ANSI_SGR << '\'';
	std::cerr << ' ';
	::error(args...);
}

inline std::optional<std::string_view> env(const std::string& key) {
	const char* value{std::getenv(key.c_str())};
	return value == nullptr ?
			std::optional<std::string_view>{std::nullopt} :
			std::optional<std::string_view>{std::string_view{value}};
}

inline std::filesystem::path tool(const std::filesystem::path& java_home, const char* const* argv) {
	const std::filesystem::path tool{argv[0]};
	return java_home / BIN / tool.filename();
}

inline void exec(const std::filesystem::path& tool, const int argc, const char* const* argv) {
	std::vector<const char*> args{};
	args.reserve(static_cast<std::size_t>(argc) + 1U);

	args.push_back(tool.c_str());
	for (int i = 1; i < argc; ++i) {
		args.push_back(argv[static_cast<std::size_t>(i)]);
	}
	args.push_back(nullptr);

	// https://stackoverflow.com/questions/190184/execv-and-const-ness
	if (::execv(tool.c_str(), const_cast<char* const*>(args.data())) == -1) {
		::java_home_error("could not be executed", ':', std::strerror(errno));
	} else {
		::error("Unknown error occurred");
	}
}

int main(const int argc, const char* const* argv) {
	const std::optional<std::string_view> java_home{::env(::JAVA_HOME)};
	if (!java_home.has_value()) {
		::java_home_error("is not defined");
		return -1;
	}

	const std::filesystem::path jdk{java_home.value()};
	if (!std::filesystem::exists(jdk)) {
		::java_home_error("does not exist");
		return -2;
	}

	const std::filesystem::path tool{::tool(jdk, argv)};
	if (!std::filesystem::exists(tool)) {
		::java_home_error("is not a valid JDK");
		return -3;
	}

	::exec(tool, argc, argv);

	return -4;
}

File: src/tool.rs
use std::borrow::Cow;
use std::borrow::Cow::{Borrowed, Owned};
use std::ffi::{c_char, CStr, CString, OsStr, OsString};
use std::io::Error;
use std::path::Path;
use std::process::ExitCode;

const ANSI_CSI: &str = "\x1b[";
const ANSI_SEPARATOR: char = ';';
const ANSI_SGR: char = 'm';
const ANSI_RESET: &str = "0";
const ANSI_BOLD: &str = "1";
const ANSI_RED: &str = "31";

const JAVA_HOME: &str = "JAVA_HOME";
const BIN: &str = "bin";

macro_rules! error {
    ( $( $arg:expr ),+ ) => {
        {
            eprint!("{ANSI_CSI}{ANSI_RED}{ANSI_SGR}");
            let mut _first: bool = true;
            $(
                if !_first {
                    eprint!(" ");
                }
                eprint!("{}", $arg);
                _first = false;
            )*
            eprint!("{ANSI_CSI}{ANSI_RESET}{ANSI_SGR}");
            eprintln!();
        }
    };
}

macro_rules! java_home_error {
    ( $( $arg:expr ),+ ) => {
        {
            eprint!("{ANSI_CSI}{ANSI_RED}{ANSI_SGR}'");
            eprint!("{ANSI_CSI}{ANSI_BOLD}{ANSI_SEPARATOR}{ANSI_RED}{ANSI_SGR}{JAVA_HOME}");
            eprint!("{ANSI_CSI}{ANSI_RED}{ANSI_SGR}'");
            eprint!(" ");
            error!( $( $arg ),+ );
        }
    };
}

macro_rules! c_string {
    ( $os_str:expr ) => {{
        match CString::new($os_str.as_encoded_bytes()) {
            Ok(os_str) => os_str,
            Err(error) => {
                java_home_error!("could not be executed", ":", error);
                return;
            }
        }
    }};
}

#[inline]
fn tool<T>(java_home: &Path, args: &[T]) -> Option<OsString>
where
    T: AsRef<OsStr>,
{
    let tool: &Path = Path::new(args.first()?);
    Some(java_home.join(BIN).join(tool.file_name()?).into_os_string())
}

#[inline]
fn execv<T>(tool: &OsStr, args: &[T])
where
    T: AsRef<OsStr>,
{
    let tool: Box<CStr> = c_string!(tool).into_boxed_c_str();
    let args: Box<[Cow<CStr>]> = {
        let mut argv: Vec<Cow<CStr>> = Vec::with_capacity(args.len());
        argv.push(Borrowed(&tool));
        for arg in &args[1..] {
            argv.push(Owned(c_string!(arg.as_ref())));
        }
        argv.into_boxed_slice()
    };

    let prog: *const c_char = tool.as_ptr();
    let argv: Box<[*const c_char]> = args
        .iter()
        .map(Cow::as_ref)
        .map(CStr::as_ptr)
        .chain([std::ptr::null()])
        .collect();

    if unsafe { libc::execv(prog, argv.as_ptr()) } == -1 {
        java_home_error!("could not be executed", ":", Error::last_os_error());
    } else {
        error!("Unknown error occurred");
    }
}

fn main() -> ExitCode {
    let args: Box<[Box<OsStr>]> = std::env::args_os()
        .map(OsString::into_boxed_os_str)
        .collect();

    let java_home: Box<OsStr> = match std::env::var_os(JAVA_HOME) {
        Some(var) => var.into_boxed_os_str(),
        None => {
            java_home_error!("is not defined");
            return ExitCode::FAILURE;
        }
    };

    let jdk: &Path = Path::new(&java_home);
    if !jdk.exists() {
        java_home_error!("does not exist");
        return ExitCode::FAILURE;
    }

    let tool: Box<OsStr> = match tool(&jdk, &args) {
        Some(tool) => tool.into_boxed_os_str(),
        None => {
            java_home_error!("is not a valid JDK");
            return ExitCode::FAILURE;
        }
    };

    execv(&tool, &args);

    ExitCode::FAILURE
}
